### Para la seccion de extracion de datos se importan las librerias necesarias.

In [26]:

# Imports
import xlrd 
import snscrape.modules.twitter as sntwitter
import pandas as pd
from datetime import datetime

### Se prosigue a   crear una lista de palabras para luego crear una lista vacia  para cada palabras

In [ ]:
frases=input()
L=[]
while frases!="exit":
    L.append(frases)
    if frases=="exit":
        break
    frases = input()
print(L)

In [ ]:
# Declaramos el diccionario
def crear_data(x):
    tabla = dict()    
    if type(x)==list:
        
        for i in range(len(x)):
    
            tabla['tweets_list'+str(i+1)] = []
    else:
       for i in range(len(x)):
            tabla['tweets_df'+str(i+1)]  = pd.DataFrame(x['tweets_list'+str(i+1)], columns=['Datetime', 'Tweet Id', 'Text', 'Username'])         
    return(tabla)

In [ ]:
tablass=crear_data(L)

In [3]:
palabras=['"Todo seria mejor si no estuviera"', '"Estoy muerto en vida"', "Solo sirvo para hacerme daño", "depresion", "ansiedad",
         "Todos mis dias son iguales", "nada tiene sentido", " MALDITA  VIDA ASQUEROSA", "Como desearia estar muerto", "me canse de escuchar esto"
         ,"Auxilio, todo lo que hago es para pedir ayuda", "maldita gente", "Odio despertar", "Solo quiero dormir", "ya me canse de todo"]
tablas = dict() # Declaramos el diccionario

for i in range(len(palabras)):
    tablas['tweets_list'+str(i+1)] = []
tablas['tweets_list4'] # para mostrar lo guardado en tabla1

[]

In [7]:
# establecemos el numero de tweets que deseamos extraer
maxTweets = 100
start_time = datetime.now()
# Se crea una lista para rellenarla con los datos extraidos de la web


#se usa  hace un bucle para hacer el webscrapping con cada palabra y que se concatene en la lista vacia
for j in range(len(tablass)):
    for i,tweet in enumerate(sntwitter.TwitterSearchScraper(L[j]).get_items()):
        if i>maxTweets:
            break
        tablass['tweets_list'+str(j+1)].append([tweet.date, tweet.id, tweet.content, tweet.user.username])
end_time = datetime.now()
#Se hace un contador para saber el tiempo que se demora el algoritmo en extraer todos los datos
print('Duration: {}'.format(end_time - start_time))      

Duration: 0:31:25.968602


for j in range(0,1):
    print(j+1)

In [ ]:
df=crear_data(tablass)

In [8]:
# se crea un dataframe para cada lista de datos 
dataframe= dict()
for i in range(len(tablas)):
    dataframe['tweets_df'+str(i+1)]  = pd.DataFrame(tablas['tweets_list'+str(i+1)], columns=['Datetime', 'Tweet Id', 'Text', 'Username'])




In [10]:
dataframe['tweets_df1'].head() ### se muestran los primeros 5 datos del primer dataframe

Datetime             Tweet Id  \
0 2022-09-24 01:47:39+00:00  1573489307020677120   
1 2022-09-23 22:54:17+00:00  1573445676960452609   
2 2022-09-18 12:02:15+00:00  1571469649258201088   
3 2022-09-17 06:33:26+00:00  1571024513529450499   
4 2022-09-11 21:48:36+00:00  1569080492498030592   

                                                Text        Username  
0           todo sería mejor si no estuviera mi papá   aAdriana_Duve  
1     siento q todo sería mejor si no estuviera aquí       itselenxx  
2  @beslindura todo sería mejor si no estuviera c...  Kim_mipatrona1  
3                Todo sería mejor si no estuviera yo   LesslieRojas1  
4  todo seria mejor si no estuviera a 2 faltas de...    joacoobrandi

### Se procede a eliminar la columna username para mantener el anonimato de los usuarios y ademas se desecha la columna Tweet Id debido a que es innecesaria para el analisis

In [11]:
for i in range(len(dataframe)):
     dataframe['tweets_df'+str(i+1)]=dataframe['tweets_df'+str(i+1)].drop(['Tweet Id', 'Username'], axis=1)

In [12]:
dataframe['tweets_df1'].head() 

Datetime                                               Text
0 2022-09-24 01:47:39+00:00           todo sería mejor si no estuviera mi papá
1 2022-09-23 22:54:17+00:00     siento q todo sería mejor si no estuviera aquí
2 2022-09-18 12:02:15+00:00  @beslindura todo sería mejor si no estuviera c...
3 2022-09-17 06:33:26+00:00                Todo sería mejor si no estuviera yo
4 2022-09-11 21:48:36+00:00  todo seria mejor si no estuviera a 2 faltas de...

### Se eliminaron las columnas deseadas

### Se unen todos los datos en un unico dataframe.

In [13]:
df=pd.concat(dataframe, ignore_index=True)

In [14]:
df.head()

Datetime                                               Text
0 2022-09-24 01:47:39+00:00           todo sería mejor si no estuviera mi papá
1 2022-09-23 22:54:17+00:00     siento q todo sería mejor si no estuviera aquí
2 2022-09-18 12:02:15+00:00  @beslindura todo sería mejor si no estuviera c...
3 2022-09-17 06:33:26+00:00                Todo sería mejor si no estuviera yo
4 2022-09-11 21:48:36+00:00  todo seria mejor si no estuviera a 2 faltas de...

In [71]:
df

Datetime  \
0     2022-09-24 01:47:39+00:00   
1     2022-09-23 22:54:17+00:00   
2     2022-09-18 12:02:15+00:00   
3     2022-09-17 06:33:26+00:00   
4     2022-09-11 21:48:36+00:00   
...                         ...   
67331 2022-07-04 11:44:37+00:00   
67332 2022-07-04 11:33:33+00:00   
67333 2022-07-04 11:33:27+00:00   
67334 2022-07-04 11:18:44+00:00   
67335 2022-07-04 10:24:53+00:00   

                                                    Text  
0               todo sería mejor si no estuviera mi papá  
1         siento q todo sería mejor si no estuviera aquí  
2      @beslindura todo sería mejor si no estuviera c...  
3                    Todo sería mejor si no estuviera yo  
4      todo seria mejor si no estuviera a 2 faltas de...  
...                                                  ...  
67331  Aaaaa csm todo, ya me canse de estar tiste, as...  
67332  Después de llorar 5 hs seguidas me di cuenta q...  
67333  Ya me canse de fingir que todo esta bien, cuan...  
67334  Nunca conocí alguien así \nQue me hiciera olvi...  
67335  Voy a decir una cosa, estoy haciendo una limpi...  

[67336 rows x 2 columns]

### Tenemos 67336 observacions en total de todo el proceso de extracción

# CLASIFICACIÓN DE LOS DATOS

### Importamos el transformer robertuito 

In [23]:
from transformers import AutoTokenizer, TFAutoModelForSequenceClassification

tokenizer = AutoTokenizer.from_pretrained("pysentimiento/robertuito-sentiment-analysis")

model = TFAutoModelForSequenceClassification.from_pretrained("pysentimiento/robertuito-sentiment-analysis")

Downloading:   0%|          | 0.00/925 [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/435M [00:00<?, ?B/s]

All model checkpoint layers were used when initializing TFRobertaForSequenceClassification.

All the layers of TFRobertaForSequenceClassification were initialized from the model checkpoint at pysentimiento/robertuito-sentiment-analysis.
If your task is similar to the task the model of the checkpoint was trained on, you can already use TFRobertaForSequenceClassification for predictions without further training.


In [33]:
classifier = pipeline("sentiment-analysis", model=model, tokenizer=tokenizer) ### se establece la función de clasificación

### Debido a la cantidad de datos se hara primero una prueba solo con 10000 datos y ver el desempeño del modelo

In [46]:
df1 = df.head(10000)

###  Ejecutamos el modelo para que clasifique en los tres posibles valores (negativo, positivo y neutral) las cadenas de texto que hay en el dataset

In [53]:
df1 = (df1.assign(sentiment = lambda x: x['Text'].apply(lambda s: classifier(s)))
.assign(
label = lambda x: x['sentiment'].apply(lambda s: (s[0]['label'])),
score = lambda x: x['sentiment'].apply(lambda s: (s[0]['score']))))
df1.head()

Datetime  \
0 2022-09-24 01:47:39+00:00   
1 2022-09-23 22:54:17+00:00   
2 2022-09-18 12:02:15+00:00   
3 2022-09-17 06:33:26+00:00   
4 2022-09-11 21:48:36+00:00   

                                                Text  \
0           todo sería mejor si no estuviera mi papá   
1     siento q todo sería mejor si no estuviera aquí   
2  @beslindura todo sería mejor si no estuviera c...   
3                Todo sería mejor si no estuviera yo   
4  todo seria mejor si no estuviera a 2 faltas de...   

                                         sentiment label     score  
0  [{'label': 'NEG', 'score': 0.9858847260475159}]   NEG  0.985885  
1  [{'label': 'NEG', 'score': 0.8769884705543518}]   NEG  0.876988  
2  [{'label': 'NEU', 'score': 0.6964797377586365}]   NEU  0.696480  
3  [{'label': 'NEG', 'score': 0.9781194925308228}]   NEG  0.978119  
4  [{'label': 'NEG', 'score': 0.9653470516204834}]   NEG  0.965347

### HACEMOS UNA PRUEBA DE LA PRIMERA COLUMNA PARA VER EL RESULTADO

In [78]:
print(df1[ 'Text'][0])
print(df1[ 'label'][0])

todo sería mejor si no estuviera mi papá
NEG


### Parece ser que  la clasificacion negativa corresponde con el contexto de la frase

In [68]:
import plotly.express as px
fig = px.pie(df1, values='score', names='label', title='Clasficación ')
fig.show()

### De los 1000 datos el 83.8 % tienen sentimientos negativos, el 14% son neutrales y tan solo el 2.16% son positivos